New serverless pattern - apigw-edge-custom-domain (#2112)
aws-samples/serverless-patterns: apigw-edge-custom-domain/README.md

> # AWS API Gateway @EDGE with custom domain name

Reviewer: ```suggestion
# Edge-optimized AWS API Gateway with custom domain name
```

> This pattern provides a simple stack (API Gateway /hello) to deploy an EDGE-Optimized API Gateway in your preferred region with ACM certificate in us-east-1(mandatory with EDGE API).

Reviewer: ```suggestion
This pattern deploys an Edge-optimized API Gateway with a single method (/hello) in your preferred region with an AWS Certificate Manager certificate in us-east-1 (which is mandatory for Edge-optimized API Gateways).
```

## Requirements
* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
* [AWS Serverless Application Model](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) (AWS SAM) installed
* Having your Route53 zone in the same AWS account.

## Deployment Instructions
1. Create the certificate in us-east-1
    ```

>     $ sam deploy -t certificate.yaml --stack-name cert-sample-api --parameter-overrides "Domain=sample.com" --region us-east-1 --resolve-s3 --capabilities CAPABILITY_IAM --no-fail-on-empty-changeset --no-progressbar

Reviewer: see above

    ```
1. Create the main stack in your preferred region
    ```
    $ sam deploy -t template.yaml --stack-name sample-api --parameter-overrides "Env=Dev Domain=sample.com" --region eu-west-1 --resolve-s3 --capabilities CAPABILITY_NAMED_IAM --no-fail-on-empty-changeset --no-progressbar
    ```
## How it works

> <img width="1273" alt="image" src="assets/architecture.png">

Reviewer: Please also provide a textual description on how the pattern works

Reviewer: Especially with the usage of custom resources

## Testing

>     $ wget sample-api.<your-domain-name>/hello

Reviewer: Please provide some context and also how the user can get the domain name, e.g. output of a CloudFormation stack

## Cleanup

1. Delete the stack
    ```
    $ sam delete --stack-name sample-api --region eu-west-1
    $ sam delete --stack-name cert-sample-api --region us-east-1
    ```

----

> Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.

Reviewer: ```suggestion
Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

        --no-progressbar # Do not display a progress bar when uploading artifacts to Amazon S3
    ```
1. Create the main stack in your preferred region
    ```
    $ sam deploy \ # Deploy command for SAM
        -t template.yaml \ # the path to the SAM template to deploy our application (EDGE API Gateway)
        --stack-name sample-api \ # Specifies the name of the CloudFormation stack
        --parameter-overrides "Domain=sample.com" \ # Overrides parameters defined in the SAM template
        --region eu-west-1 \ # Specifies your preferred AWS region where the stack will be deployed, eu-west-1 for me
        --resolve-s3 \ # Automatically create an Amazon S3 bucket to use for packaging and deploying
        --capabilities CAPABILITY_IAM \ # Specifies the IAM capabilities required for the stack to create/modify IAM resources
        --no-fail-on-empty-changeset \ # Indicates that the deployment should not fail if there are no changes to be made
        --no-progressbar # Do not display a progress bar when uploading artifacts to Amazon S3
    ```
## How it works
<img width="1273" alt="image" src="assets/architecture.png">

> In this pattern we use CloudFormation custom resources, you can check out this [workshop](https://mng.workshop.aws/cloudformation/customresource.html).

Reviewer: Please include a small blurb like "resources which allow you to [...] For more information, please refer to this [...]"

Author: done

1. As you can see in the diagram above, the first step is to create the "certificate.yaml" file.

> To create an ACM certificate on AWS with DNS as DomainValidationOptions, we need to provide the DomainName and the HostedZoneId. CloudFormation does not get the HostedZoneId dynamically from now. I made a custom resource for a dynamic approach, not hardcoded.

Reviewer: ```suggestion
To create an ACM certificate on AWS with DNS as ´DomainValidationOptions´, we need to provide the ´DomainName´ and the ´HostedZoneId´. CloudFormation does not get the ´HostedZoneId´ dynamically for now. A custom resource is used to get these values dynamically instead of hardcoding them.
```

Author: done
